policy/release/sbom_cyclonedx/sbom_cyclonedx.rego

```diff
@@ -148,6 +148,63 @@ deny contains result if {
 	result := lib.result_helper_with_term(rego.metadata.chain(), [id, reference.url, reference.type, msg], id)
 }
 
+# METADATA
+# title: Allowed package sources
+# description: >-
+#   Confirm the CycloneDX SBOM packages fetched by Cachi2 that define externalReferences
+#   of type distribution are only those that satisfy any of the patterns defined for the
+#   given purl type. By default no sources are allowed. Use the "allowed_package_sources"
+#   rule data key to provide a list of patterns for each purl type.
+# custom:
+#   short_name: allowed_package_sources
+#   failure_msg: Package %s fetched by cachi2 was sourced from %q which is not allowed
+#   solution: Update the image to not use a package from a disallowed source.
+#   collections:
+#   - redhat
+#   - policy_data
+#   effective_on: 2024-11-30T00:00:00Z
+deny contains result if {
+	some s in sbom.cyclonedx_sboms
+	some component in s.components
+	some reference in component.externalReferences
+	some properties in component.properties
+
+	parsed_purl := ec.purl.parse(component.purl)
+
+	# only look at components fetched by cachi2 that define an externalReferences
```

Comment:
I do wonder if this rule can be a bit more data driven. For example, what if we allowed the rule data to be something like this:

```yaml
- type: maven
  properties:
    - name: cachi2:found_by
      value: cachi2
  reference:
    type: distribution
  allowed_references:
    - ".*apache.org.*"
    - ".*example.com.*"
```

Or some variation of that. WDYT?

Reply:
(Also I think we should settle this first before I get to some of the other comments)

Reply:
@lcarva @simonbaird any thoughts on this?

Reply:
Let's keep it as is. We can always evolve it later on if the need becomes obvious.